### PR TITLE
[declare] Merge DeclareDef into Declare

### PR DESCRIPTION
**Pull request by ejgallego**
The API in `DeclareDef` should become the recommended API in `Declare`.

This greatly reduces the exposure of internals; we still have a large
offender in `Lemmas` but that will be taken care of in the next
commit; effectively removing quite some chunks from `declare.mli`.

The PR refactoring `Lemmas` is not included here as we want to merge #11836 first [after we fix it to support Equations chain obligation solving inside hooks]

This depends on #12236 tho they could be merged together.

### CODE REVIEW

**Comment by ejgallego** [comment]
> This depends on #12236 tho they could be merged together.

This is also ready, I guess #12236 is too small so let's make it a part of this one.

**Review by Matafou — APPROVED** [review]

File: vernac/declare.mli [hunk 1]
```diff
@@ -206,16 +206,12 @@ exception AlreadyDeclared of (string option * Id.t)
...
 module Internal : sig
 
-  val map_entry_body : f:('a Entries.proof_output -> 'b Entries.proof_output) -> 'a proof_entry -> 'b proof_entry
-  val map_entry_type : f:(Constr.t option -> Constr.t option) -> 'a proof_entry -> 'a proof_entry
-  (* Overriding opacity is indeed really hacky *)
-  val set_opacity : opaque:bool -> 'a proof_entry -> 'a proof_entry
+   val map_entry_body : f:('a Entries.proof_output -> 'b Entries.proof_output) -> 'a proof_entry -> 'b proof_entry
+   val map_entry_type : f:(Constr.t option -> Constr.t option) -> 'a proof_entry -> 'a proof_entry
+   (* Overriding opacity is indeed really hacky *)
+   val set_opacity : opaque:bool -> 'a proof_entry -> 'a proof_entry
```
Comment (SkySkimmer) [inline vernac/declare.mli:207]: Spurious change?

File: vernac/declareUniv.ml [hunk 2]
```diff
@@ -10,6 +10,17 @@
 
 open Names
 
+(* object_kind , id *)
```
Comment (SkySkimmer) [inline vernac/declareUniv.ml:13]: This comment should rather be in the mli

File: vernac/vernacexpr.ml [hunk 3]
```diff
@@ -286,6 +288,22 @@ type extend_name =
...
+
+type hints_expr =
+  | HintsResolve of (hint_info_expr * bool * reference_or_constr) list
+  | HintsResolveIFF of bool * Libnames.qualid list * int option
+  | HintsImmediate of reference_or_constr list
+  | HintsUnfold of Libnames.qualid list
+  | HintsTransparency of Libnames.qualid Hints.hints_transparency_target * bool
+  | HintsMode of Libnames.qualid * Hints.hint_mode list
+  | HintsConstructors of Libnames.qualid list
+  | HintsExtern of int * Constrexpr.constr_expr option * Genarg.raw_generic_argument
```
Comment (SkySkimmer) [inline vernac/vernacexpr.ml:305]: Why did this stuff move? Maybe copy the explanation to the commit message.

**Review by SkySkimmer — APPROVED** [review]

File: vernac/declare.mli (hunk 1 above)
Comment (ejgallego) [reply vernac/declare.mli:207]: Yes, sorry. Fixed. [originally due to a very different internal removal strategy]

File: vernac/declareUniv.ml (hunk 2 above)
Comment (ejgallego) [reply vernac/declareUniv.ml:13]: Done, thanks [I'd like to get rid of this anyways]

File: vernac/vernacexpr.ml (hunk 3 above)
Comment (ejgallego) [reply vernac/vernacexpr.ml:305]: Actually I had to scratch my head quite a bit with this; the dependency cycle was:
- `Declare: Vernacexpr.decl_notation list`
- `Vernacexpr: ComHint.hint_expr`
- `ComHint`: `Declare.declare_constant`

There are quite a few ways to solve this, for example by adding the AST to `Metasyntax` , having `Declare` not to depend on `Vernacexpr` [which seems actually sane in the medium term], reworking notation support.... I chose moving hints ast to vernacexpr as the approach least invasive and more consistent with the current style, but indeed we may want implement some other solution in the medium term.

Comment added to the commit message.

**Pull request merged**